content/blog/2018/delayering-istio/delayering-istio.md

```diff
@@ -0,0 +1,216 @@
+
+### Artifacts of Container Networking
+
+Extending network connectivity from host into the container has been a [major challenge](https://kubernetes.io/blog/2016/01/why-kubernetes-doesnt-use-libnetwork/).  New layers of network plumbing were invented explicitly for that purpose.  As such, an application running in a container is simply a process on the host.  However due to a [fundamental misalignment](http://appswitch.io/blog/kubernetes_istio_and_network_function_devirtualization_with_appswitch/) between the network abstraction expected by the application and the abstraction exposed by container network namespace, the process cannot directly access the host network.  Applications think of networking in terms of sockets or sessions whereas network namespaces expose a device abstraction.  Once placed in a network namespace, the process suddenly loses all connectivity.  The notion of veth-pair and corresponding tooling were invented just to close that gap.  The data would now have to go from a host interface into a virtual switch and then through a veth-pair to the virtual network interface of the container network namespace.
+
+AppSwitch can effectively remove both the virtual switch and veth-pair layers on both ends of the connection.  Since the connections are established by the daemon running on the host using the network that’s already available on the host, there is no need for additional plumbing to bridge host network into the container.  The socket FDs created on the host are passed to the application running within the pod’s network namespace.  By the time the application receives the FD, all control path work (security checks, connection establishment) is already done and the FD is ready for actual IO.
+
+### Skip TCP/IP for colocated endpoints
+
+TCP/IP is the universal protocol medium over which pretty much all communication occurs.  But if application endpoints happen to be on the same host, is TCP/IP really required?  After all, it does do quite a bit of work and it is quite complex.  Unix sockets are explicitly designed for intra-host communication and AppSwitch can transparently switch the communication to occur over a Unix socket for colocated endpoints.
+
+For each listening socket of an application, AppSwitch maintains two listening sockets, one each for TCP and Unix.  When a client tries to connect to a server that happens to be colocated, AppSwitch daemon would choose to connect to the Unix listening socket of the server.  The resulting Unix sockets on each end are passed into respective applications.  Once a fully connected FD is returned, the application would simply treat it as a bitpipe.  The protocol doesn’t really matter.  The application may occasionally make protocol specific calls such as `getsockname(2)` and AppSwitch would handle them in kind.  It would present consistent responses such that the application would continue to run on.
+
```

Comment:
It would be nice to compare/contrast with the recent work in the kernel that eliminated the need to go through the TCP/IP protocol stack for loopback socket communication. https://www.slideshare.net/ThomasGraf5/accelerating-envoy-and-istio-with-cilium-and-the-linux-kernel

Reply:
Yes that would be a good, but somewhat out of scope for this blog.  Perhaps a follow up...